lib/cloud/azure/redis_enterprise.go

```diff
@@ -102,6 +116,116 @@ func (c *redisEnterpriseClient) getClusterAndDatabaseName(id *arm.ResourceID) (s
 	}
 }
 
-// RedisEnterpriseClusterDefaultDatabase is the default database name for a
-// Redis Enterprise cluster.
-const RedisEnterpriseClusterDefaultDatabase = "default"
+// ListAll returns all Azure Redis Enterprise databases within an Azure subscription.
+func (c *redisEnterpriseClient) ListAll(ctx context.Context) ([]*RedisEnterpriseDatabase, error) {
+	var allDatabases []*RedisEnterpriseDatabase
+	pager := c.clusterAPI.NewListPager(&armredisenterprise.ClientListOptions{})
+	for pageNum := 0; pager.More(); pageNum++ {
+		page, err := pager.NextPage(ctx)
+		if err != nil {
+			return nil, trace.Wrap(ConvertResponseError(err))
+		}
+
+		databases, err := c.listDatabasesByClusters(ctx, page.Value)
+		if err != nil {
+			return nil, trace.Wrap(ConvertResponseError(err))
+		}
+		allDatabases = append(allDatabases, databases...)
+	}
+	return allDatabases, nil
+}
+
+// ListWithinGroup returns all Azure Redis Enterprise databases within an Azure resource group.
+func (c *redisEnterpriseClient) ListWithinGroup(ctx context.Context, group string) ([]*RedisEnterpriseDatabase, error) {
+	var allDatabases []*RedisEnterpriseDatabase
+	pager := c.clusterAPI.NewListByResourceGroupPager(group, &armredisenterprise.ClientListByResourceGroupOptions{})
+	for pageNum := 0; pager.More(); pageNum++ {
+		page, err := pager.NextPage(ctx)
+		if err != nil {
+			return nil, trace.Wrap(ConvertResponseError(err))
+		}
+
+		databases, err := c.listDatabasesByClusters(ctx, page.Value)
+		if err != nil {
+			return nil, trace.Wrap(ConvertResponseError(err))
+		}
+		allDatabases = append(allDatabases, databases...)
+	}
+	return allDatabases, nil
+}
+
+// listDatabasesByClusters fetches databases for the provided clusters.
+func (c *redisEnterpriseClient) listDatabasesByClusters(ctx context.Context, clusters []*armredisenterprise.Cluster) ([]*RedisEnterpriseDatabase, error) {
+	var allDatabases []*RedisEnterpriseDatabase
+	for _, cluster := range clusters {
+		if cluster == nil { // should never happen, but checking just in case.
+			continue
+		}
+
+		// If listDatabasesByCluster fails for any reason, make a log and continue to
+		// other clusters.
+		databases, err := c.listDatabasesByCluster(ctx, cluster)
+		if err != nil {
+			if trace.IsAccessDenied(err) || trace.IsNotFound(err) {
```

Comment:
Question: Reporting the access denied error as a warning log wouldn't be better for users configuring the auto-discovery. I'm seeing this error as a misconfiguration (lack of Azure permissions), so if users get this message as a warning, they can double-check if their permissions are correctly set.

Note: I see possible log spam when users have a large set of databases we don't have access to. But it might be helpful anyway.

Reply:
Besides spamming, one of the reasons we use debug is that the agent may have permission to some resources but not others. This can be intentional by the user to limit our permissions.

Even though that reason doesn't apply to this case, I would still keep it consistent with other places (as user has to turn on debugging to see all other permssion problems anyway)